File: Web2Projects/apis/Python/stamping/atestaciones.py
import requests
import hashlib
import base64
import json

# Definimos el contenido del documento
credential_subject = {
    "id": "did:dni:pe:10203040",
    "type": "Constancia de trabajo",
    "subjectName": "Juan Perez",
    "role": "Jefe de Planta",
    "startOfWork": "01/07/2018"
}

# Paso 2: Calcular hash SHA256
json_string = json.dumps(credential_subject, separators=(',', ':'), ensure_ascii=False)
evidence_hash = hashlib.sha256(json_string.encode('utf-8')).hexdigest()
base64_data = base64.b64encode(json_string.encode('utf-8')).decode()

# Paso 4: Llamada al API de Stamping.io
# Token proporcionado por Stamping.io
API_TOKEN = "98cdb8c28f159e8021b736173b9dbe8fa1d8a6d4bc68bf40f6448b05d7a"

# Construir URL
url = f"https://api.stamping.io/stamp/?evidence={evidence_hash}&data={base64_data}&async=true&token={API_TOKEN}"

# Ejecutar la petición
response = requests.get(url)

print("Status:", response.status_code)
print("Respuesta:", response.json())

File: Web2Projects/apis/java/stamping/atestaciones.java
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.security.MessageDigest;
import java.util.Base64;

public class atestaciones {

    public static void main(String[] args) throws Exception {
        // 1. Definir la afirmación en formato JSON (sin espacios ni saltos innecesarios)
        String json = "{\"id\":\"did:dni:pe:10203040\",\"type\":\"Constancia de trabajo\",\"subjectName\":\"Juan Perez\",\"role\":\"Jefe de Planta\",\"startOfWork\":\"01/07/2018\"}";

        // 2. Calcular SHA256 hash del JSON
        MessageDigest digest = MessageDigest.getInstance("SHA-256");
        byte[] hashBytes = digest.digest(json.getBytes("UTF-8"));
        StringBuilder hexHash = new StringBuilder();
        for (byte b : hashBytes) {
            hexHash.append(String.format("%02x", b));
        }
        String evidenceHash = hexHash.toString();

        // 3. Codificar JSON en Base64
        String base64Data = Base64.getEncoder().encodeToString(json.getBytes("UTF-8"));

        // 4. Token de acceso proporcionado por Stamping.io
        String token = "98cdb8c28f159e8021b736173b9dbe8fa1d8a6d4bc68bf40f6448b05d7a";

        // 5. Construir la URL de la API
        String url = String.format(
            "https://api.stamping.io/stamp/?evidence=%s&data=%s&async=true&token=%s",
            evidenceHash, base64Data, token
        );

        // 6. Crear cliente HTTP y solicitud
        HttpClient client = HttpClient.newHttpClient();
        HttpRequest request = HttpRequest.newBuilder()
            .uri(URI.create(url))
            .GET()
            .build();

        // 7. Enviar solicitud y mostrar respuesta
        HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());

        System.out.println("Status: " + response.statusCode());
        System.out.println("Respuesta: " + response.body());
    }
}
